### ISSUE TITLE
Question about the accuracy of pretrained FCN model

### ISSUE THREAD

**Issue by ChandayHunt** [AlvinWen428/FeatureCP#1]
Hello authors,
I do have a question related to the pretrained FCN model on cityscapes. Basically I tried to load the checkpoint model provided, and apply them on the 'valid dataset' of cityscapes. The mIOU I got was less than 1%. I have noticed the `output_post_process` in `FCN` has dropped the 0th (background), and I didn't use that. Also I didn't use the `view()` function. Below are the label and prediction selected. Please take a look, thank you!

![image](https://user-images.githubusercontent.com/129140092/230742922-0e0783e3-83ee-4069-99a9-c46a3f033e0f.png)
![image](https://user-images.githubusercontent.com/129140092/230742927-1799fce0-be2d-470d-879e-32d11b9a6f86.png)

**Comment by AlvinWen428** [comment]
@ChandayHunt Thanks for your interest in our work. 

As introduced in Section 5.1 and Appendix B.1 in our paper, we transform the original pixel-wise classification problem into a high-dimensional pixel-wise regression problem for the convenience of conformal prediction. Thus, we convert the classification label space into a continuous space by the double log trick, i.e., $\dot{y}=\log(-\log(\tilde{y}))$. And we use the WeightedMSE (https://github.com/AlvinWen428/FeatureCP/blob/b1b779b991ff3107e543278052bbecea0c3e64c6/conformal/utils.py#L30) as our loss function. 

We do the same transformation when training FCN on cityscapes. If you want to evaluate the mIoU metric, you need to convert the output of the model back to the original space by `output = torch.exp(-torch.exp(output))`.  The mIoU of the provided checkpoint is 0.5798.

I have updated the README to make a clarification about this question.

**Comment by ChandayHunt** [comment]
Great, that helps a lot. I managed to draw the prediction as below, which makes sense.
![image](https://user-images.githubusercontent.com/129140092/230790352-b6d6355b-d6e6-4b6a-b6ee-47e41d4840cd.png)

**Comment by ChandayHunt** [comment]
Sorry, just the last question. I converted back to label space by `output = torch.exp(-torch.exp(output))`, however the mIOU I got was just ~50%. Could you please help me check it? Here are the codes:
```
class ConfMatrix(object):
    def __init__(self, num_classes):
        self.num_classes = num_classes
        self.mat = None

    def update(self, pred, target):
        n = self.num_classes
        if self.mat is None:
            self.mat = torch.zeros((n, n), dtype=torch.int64, device=pred.device)
        with torch.no_grad():
            k = (target >= 0) & (target < n)
            inds = n * target[k].to(torch.int64) + pred[k]
            self.mat += torch.bincount(inds, minlength=n ** 2).reshape(n, n)

    def get_metrics(self):
        h = self.mat.float()
        acc = torch.diag(h).sum() / h.sum()
        iu = torch.diag(h) / (h.sum(1) + h.sum(0) - torch.diag(h))
        return torch.mean(iu).item(), acc.item()
```

basically I set `conf_mat = ConfMatrix(num_classes=20)` and during each of the item in val loader, we use `conf_mat.update((output).flatten(), (test_label).flatten())`, where `output` and `test_label` both contain the index of class (within the range(0~19)). At last, I use `iou, acc = conf_mat.get_metrics()` to get the final results. However the result I got was `iou, acc 0.502954363822937 0.8637312054634094`. I have a few thoughts on this, basically: 1) Could it be that because I am using the validation dataset, instead of the train/calib/test dataset split as in the code? 2) Could it be that I should only consider the class index from 1 to 19, instead of including index 0? Thank you so much!

**Comment by AlvinWen428** [comment]
Your second hypothesis is correct. We ignore class 0 (unlabeled) during training and testing, following [Pytorch-ENet](https://github.com/davidtvs/PyTorch-ENet). If you exclude index 0, I think the result will be correct.

**Comment by ChandayHunt** [comment]
Yeah, I got a result close to that, 0.5656, using the updated code that ignores the index 0:
```
class ConfMatrix(object):
    def __init__(self, num_classes):
        self.num_classes = num_classes
        self.mat = None
        self.ignore_index = 0

    def update(self, pred, target):
        n = self.num_classes
        if self.mat is None:
            self.mat = torch.zeros((n, n), dtype=torch.int64, device=pred.device)
        with torch.no_grad():
            k = (target >= 0) & (target < n)
            inds = n * target[k].to(torch.int64) + pred[k]
            self.mat += torch.bincount(inds, minlength=n ** 2).reshape(n, n)

    def get_metrics(self):
        h = self.mat.float()
        h[:, self.ignore_index] = 0
        h[self.ignore_index, :] = 0
        acc = torch.diag(h).sum() / h.sum()
        iu = torch.diag(h)[1:] / (h.sum(1) + h.sum(0) - torch.diag(h))[1:]
        return torch.mean(iu).item(), acc.item()
```
Note that since index 0 is ignored, the denominator and numerators also need to omit index 0. I used the downloaded weights and tested it on the official validation set. However it is still less than 57.98%. Could you please take a look? Thank you!

**Comment by AlvinWen428** [comment]
We use a different train-validation partition from the official strategy. Please have a look at [this link](https://github.com/AlvinWen428/FCN-Trainer/blob/a7ab1ef2b20e593d35386139ea0ab5c59c15b75e/main.py#L58) to find out how we split the dataset. The 57.98% result is calculated on our new validation set. You can also have a look at our [implementation](https://github.com/AlvinWen428/FCN-Trainer/blob/a7ab1ef2b20e593d35386139ea0ab5c59c15b75e/metric/iou.py#L85) of IoU to double-check whether your code is correct.

**Comment by AlvinWen428** [comment]
If you do not have further problems, we will close this issue.
